parser/parser_test.go

```diff
@@ -0,0 +1,1109 @@
+		is.Empty(query.Columns[1].Alias)
+		is.Equal("c", query.Columns[2].Name)
+		is.Empty(query.Columns[2].Alias)
+		is.False(query.From.Only)
+		is.Equal("test2", query.From.Table.Name)
+		is.Empty(query.From.Table.Alias)
+		is.Len(query.Joins, 1)
+		is.Equal("test4", query.Joins[0].Table.Name)
+		is.Empty(query.Joins[0].Table.Alias)
+		is.Equal(types.RightJoin, query.Joins[0].Type)
+		{
+			on, ok := query.Joins[0].Condition.(stmt.OnClause)
+			is.True(ok)
+			is.NotEmpty(on)
+			is.Equal("test4.gid", on.Left.Name)
+			is.Empty(on.Left.Alias)
+			is.Equal("test2.id", on.Right.Name)
+			is.Empty(on.Right.Alias)
+		}
+		is.True(query.Where.IsEmpty())
+		is.True(query.GroupBy.IsEmpty())
+		is.True(query.Having.IsEmpty())
+		is.True(query.OrderBy.IsEmpty())
+		is.True(query.Limit.IsEmpty())
+		is.True(query.Offset.IsEmpty())
+		is.True(query.Suffix.IsEmpty())
+
+		ctx := &types.RawContext{}
+		result.Write(ctx)
+		is.Equal(input, ctx.Query())
+	}
+	{
+		input := fmt.Sprint(
+			"SELECT a, b, c FROM test2 INNER JOIN test4 ON test4.gid = test2.id ",
+			"INNER JOIN test3 ON test4.uid = test3.id",
+		)
+		result, err := parser.Parse(input)
+		is.NoError(err)
+		is.NotNil(result)
+		is.IsType(stmt.Select{}, result)
+		query := (result).(stmt.Select)
+
+		is.True(query.Prefix.IsEmpty())
+		is.True(query.With.IsEmpty())
+		is.False(query.Distinct)
+		is.Len(query.Columns, 3)
+		is.Equal("a", query.Columns[0].Name)
+		is.Empty(query.Columns[0].Alias)
+		is.Equal("b", query.Columns[1].Name)
+		is.Empty(query.Columns[1].Alias)
+		is.Equal("c", query.Columns[2].Name)
+		is.Empty(query.Columns[2].Alias)
+		is.False(query.From.Only)
+		is.Equal("test2", query.From.Table.Name)
+		is.Empty(query.From.Table.Alias)
+		is.Len(query.Joins, 2)
+		is.Equal("test4", query.Joins[0].Table.Name)
+		is.Empty(query.Joins[0].Table.Alias)
+		is.Equal(types.InnerJoin, query.Joins[0].Type)
+		{
+			on, ok := query.Joins[0].Condition.(stmt.OnClause)
+			is.True(ok)
+			is.NotEmpty(on)
+			is.Equal("test4.gid", on.Left.Name)
+			is.Empty(on.Left.Alias)
+			is.Equal("test2.id", on.Right.Name)
+			is.Empty(on.Right.Alias)
+		}
+		is.Equal("test3", query.Joins[1].Table.Name)
+		is.Empty(query.Joins[1].Table.Alias)
+		is.Equal(types.InnerJoin, query.Joins[1].Type)
+		{
+			on, ok := query.Joins[1].Condition.(stmt.OnClause)
+			is.True(ok)
+			is.NotEmpty(on)
+			is.Equal("test4.uid", on.Left.Name)
+			is.Empty(on.Left.Alias)
+			is.Equal("test3.id", on.Right.Name)
+			is.Empty(on.Right.Alias)
+		}
+		is.True(query.Where.IsEmpty())
+		is.True(query.GroupBy.IsEmpty())
+		is.True(query.Having.IsEmpty())
+		is.True(query.OrderBy.IsEmpty())
+		is.True(query.Limit.IsEmpty())
+		is.True(query.Offset.IsEmpty())
+		is.True(query.Suffix.IsEmpty())
+
+		ctx := &types.RawContext{}
+		result.Write(ctx)
+		is.Equal(input, ctx.Query())
+	}
+	{
+		input := fmt.Sprint(
+			"SELECT a, b, c FROM test2 INNER JOIN test4 ON (test4.gid = test2.id AND test4.d = test2.d) ",
```

Comment:
it's the best case scenario, you don't retrieve parenthesis around the `join` operator everywhere